### PR TITLE
[versioning] EOS-25927: Implement ListObjectVersions API version-id-marker request parameter

### PR DESCRIPTION
**Pull request by indrajitzagade**
# Problem Statement
- Implement ListObjectVersions API version-id-marker request parameter

# Design
https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/766050414/ListObjectVersions+LLD

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by sjain09 — APPROVED** [review]

File: server/s3_list_object_versions_action.cc [hunk 1]
```diff
@@ -115,7 +115,29 @@ void S3ListObjectVersionsAction::validate_request() {
...
+    if (request_version_id_marker.empty()) {
+      s3_log(S3_LOG_DEBUG, request_id, "version-id-marker is empty\n");
+      // TODO: Add invalid argument details to the response.
+      set_s3_error("InvalidArgument");
+      send_response_to_s3_client();
+      return;
+    } else if (request_key_marker.empty()) {
+      s3_log(S3_LOG_DEBUG, request_id, "key-marker is empty\n");
+      // TODO: Add invalid argument details to the response.
+      set_s3_error("InvalidArgument");
```
Comment (keithpine) [inline server/s3_list_object_versions_action.cc:132]:
```suggestion
      set_s3_error("InvalidArgument");
      set_s3_error_message("A version-id marker cannot be specified without a key marker.");
      set_invalid_argument("version-id-marker", request_version_id_marker);
```

Here's the result from AWS:

```txt
$ aws --debug s3api list-object-versions --bucket mybucket --version-id-marker 8e5_5Mm3btVpNYmt1rQZnFfkGM0aRgd9

An error occurred (InvalidArgument) when calling the ListObjectVersions operation: A version-id marker cannot be specified without a key marker.
```

Debug output:
```xml
<?xml version="1.0" encoding="UTF-8"?>\n
<Error>
    <Code>InvalidArgument</Code>
    <Message>A version-id marker cannot be specified without a key marker.</Message>
    <ArgumentName>version-id-marker</ArgumentName>
    <ArgumentValue>8e5_5Mm3btVpNYmt1rQZnFfkGM0aRgd9</ArgumentValue>
    <RequestId>XJZATXKZAK5DZJW6</RequestId>
    <HostId>9bn+DPs/Ec90ovM8aBQlbn3oktn/YjKXxmUGKZkuVhKzVhR4uJJQRmoWxFU1SoUdB8Hll8a20oQ=</HostId>
</Error>
```

File: server/s3_list_object_versions_action.cc [hunk 2]
```diff
@@ -115,7 +115,29 @@ void S3ListObjectVersionsAction::validate_request() {
...
-  last_key = request_key_marker;  // As requested by user
+  if (request->has_query_param_key("version-id-marker")) {
+    request_version_id_marker =
+        request->get_query_string_value("version-id-marker");
+    s3_log(S3_LOG_INFO, request_id, "request_version_id_marker = %s\n",
+           request_version_id_marker.c_str());
+    if (request_version_id_marker.empty()) {
+      s3_log(S3_LOG_DEBUG, request_id, "version-id-marker is empty\n");
+      // TODO: Add invalid argument details to the response.
+      set_s3_error("InvalidArgument");
```
Comment (keithpine) [inline server/s3_list_object_versions_action.cc:126]:
```suggestion
      set_s3_error("InvalidArgument");
      set_s3_error_message("A version-id marker cannot be empty.");
      set_invalid_argument("version-id-marker", request_version_id_marker);
```

Here's the result from AWS:

```txt
$ aws s3api list-object-versions --bucket mybucket --version-id-marker '' --key-marker shell.nix

An error occurred (InvalidArgument) when calling the ListObjectVersions operation: A version-id marker cannot be empty.
```

Debug output:

```xml
<?xml version="1.0" encoding="UTF-8"?>\n
<Error>
    <Code>InvalidArgument</Code>
    <Message>A version-id marker cannot be empty.</Message>
    <ArgumentName>version-id-marker</ArgumentName>
    <ArgumentValue></ArgumentValue>
    <RequestId>VH8P4E93T962G3WX</RequestId>
    <HostId>iRlqrKmf+/CcVbcE6qjncj4SoHvZajRb9wq1n6CL7cvpXtnJMTealm20ZMwHuzMpaXjwenZ+c+k=</HostId>
</Error>
```

**Review by keithpine — APPROVED** [review]
So the behavior of CORTX ListObjectVersions for truncated listings is different from both AWS and Minio. AWS and Minio are identical.

I would suggest storing multiple keys and multiple versions, and trying a listing with `--max-keys` set yourself. You will see the response values are different.

For example, I stored 5 versions of key `shell.nix` then 2 versions of `z.bar`. If I do a listing with `--max-keys 5`, the results between AWS/Minio and CORTX are quite different. Also try using `--key-marker` and `--version-id-marker` and I think the behavior is also different. AWS seems to start from the next key and version ID, while this is starting from the key/version ID given.

Reading the API docs, I don't understand why AWS responds and behaves this way.

If you want to merge this and follow up with additional changes, that's fine by me. We should just be aware that this current behavior is not compatible with AWS yet.

Added this to https://jts.seagate.com/browse/EOS-27169.

**Comment by indrajitzagade** [comment]
> So the behavior of CORTX ListObjectVersions for truncated listings is different from both AWS and Minio. AWS and Minio are identical.
> 
> I would suggest storing multiple keys and multiple versions, and trying a listing with `--max-keys` set yourself. You will see the response values are different.
> 
> For example, I stored 5 versions of key `shell.nix` then 2 versions of `z.bar`. If I do a listing with `--max-keys 5`, the results between AWS/Minio and CORTX are quite different. Also try using `--key-marker` and `--version-id-marker` and I think the behavior is also different. AWS seems to start from the next key and version ID, while this is starting from the key/version ID given.
> 
> Reading the API docs, I don't understand why AWS responds and behaves this way.
> 
> If you want to merge this and follow up with additional changes, that's fine by me. We should just be aware that this current behavior is not compatible with AWS yet.
> 
> Added this to https://jts.seagate.com/browse/EOS-27169.

File: server/s3_list_object_versions_action.cc (hunk 1 above)
Comment (indrajitzagade) [reply server/s3_list_object_versions_action.cc]: Done

File: server/s3_list_object_versions_action.cc (hunk 2 above)
Comment (indrajitzagade) [reply server/s3_list_object_versions_action.cc]: Done

**Review by sanalpk — APPROVED** [review]

**Comment by indrajitzagade** [comment]
Build job: http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1685/ (Passed)
Deployment job: http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/680/ (Passed)
Pre-merge job: http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/955/ (In progress)

**Comment by indrajitzagade** [comment]
It was discussed and decided to take-up the failing pre-merge job issue later on the object-versioning branch, hence merged this PR.

**Pull request merged**